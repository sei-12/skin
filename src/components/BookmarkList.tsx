import { Stack } from "@mui/material"
import { IData } from "../dts/data"
import { useEffect, useRef, useState } from "react"
import { BookmarkItem } from "./BookmarkItem"

export function useBookmarkList(
	onClickRemove: (key: number) => void,
	onClickEdit: (key: number) => void,
){
    const [items, setItems] = useState<IData.Bookmark[]>([])
    const [focusIndex, setFocusIndex] = useState(0)
	const itemRefs = useRef<(HTMLDivElement | null)[]>([]);
    
	useEffect(() => {
		// フォーカスが変更されたときにスクロールする
		const focusedItem = itemRefs.current[focusIndex];
		if (focusedItem) {
			focusedItem.scrollIntoView({ block: "nearest"});
		}
	}, [focusIndex]);

    return {
        items,
        setItems,
        focusIndex,
        setFocusIndex,
        
        props: {
            onClickEdit,
            onClickRemove,
            itemRefs,
            focusIndex,
            items,
        }
    }
}

export function BookmarkList(props: ReturnType<typeof useBookmarkList>["props"]){
    return (
        <Stack 
            spacing={1}
            sx={{
                overflow: "scroll",
                '&::-webkit-scrollbar': {
                    display: 'none', 
                },
                flexGrow: 1,
                padding: 0.5
            }}
        >
            {
                props.items.map( (item,i) => {
                    return <BookmarkItem 
						ref={(el) => (props.itemRefs.current[i] = el)}
                        data={item}
                        focus={props.focusIndex === i}                        
                        key={item.id}
                        onClickEdit={props.onClickEdit}
                        onClickRemove={props.onClickRemove}
                    ></BookmarkItem>
                })                
            }
        </Stack>
    )
}