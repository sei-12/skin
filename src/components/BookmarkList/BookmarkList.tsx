
// type BookmarkListProps = {
// 	onclickremove: (key: string) => void
// 	onclickedit: (key: string) => void
// 	// items: bookmarkdata[]
//     focusindex: number
// }


// export function BookmarkList(){

// }