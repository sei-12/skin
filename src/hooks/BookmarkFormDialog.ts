import { useCallback, useState } from "react";
import type { BookmarkFormDialogProps } from "../components/BookmarkFormDialog";
import { useConfig } from "../providers/configProvider";

export function useBookmarkFormDialog(onClickDone: () => void) {
    const [open, setOpen] = useState(false);

    const onClickCancel = useCallback(() => {
        setOpen(false);
    }, []);
    const onClickDone_ = useCallback(() => {
        setOpen(false);
        onClickDone();
    }, []);

    const { colorTheme } = useConfig();
    const props: BookmarkFormDialogProps = {
        colorTheme,
        onClickCancel,
        onClickDone: onClickDone_,
        open,
    };
    return {
        props,
        setOpen,
    };
}
