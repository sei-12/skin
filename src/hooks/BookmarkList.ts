import { useCallback, useEffect, useRef, useState } from "react";
import type { IData } from "../dts/data";
import type { BookmarkListProps } from "../components/BookmarkList";

export function useBookmarkList(
    onClickRemove: (key: number) => void,
    onClickEdit: (key: number) => void
) {
    const [items, setItems] = useState<IData.Bookmark[]>([]);
    const [focusIndex, setFocusIndex] = useState(0);
    const itemRefs = useRef<(HTMLDivElement | null)[]>([]);

    useEffect(() => {
        // フォーカスが変更されたときにスクロールする
        const focusedItem = itemRefs.current[focusIndex];
        if (focusedItem) {
            focusedItem.scrollIntoView({ block: "nearest" });
        }
    }, [focusIndex]);

    const focusUp = useCallback(() => {
        setFocusIndex((cur) => {
            let newIndex = cur - 1;
            if (newIndex < 0) {
                newIndex = items.length - 1;
            }
            return newIndex;
        })
    }, [items])

    const focusDown = useCallback(() => {
        setFocusIndex((cur) => {
            let newIndex = cur + 1;
            if (newIndex >= items.length) {
                newIndex = 0;
            }
            return newIndex;
        })
    }, [items])


    const props: BookmarkListProps = {
        onClickEdit,
        onClickRemove,
        itemRefs,
        focusIndex,
        items,
    };

    return {
        items,
        setItems,
        // focusIndex,
        setFocusIndex,
        
        focusDown,
        focusUp,
        props,
    };
}