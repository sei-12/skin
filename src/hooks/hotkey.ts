import { useHotkeysContext } from "react-hotkeys-hook";

//
// TODO
// ページを分けたから必要ない
// default,suggestion_windowの二つでいいと思う
//
export const HOTKEY_SCOPES = {
    SEARCH_BOOKMARK: "search_bookmark",
    SEARCH_BOOKMARK_SUGGESTION_WINDOW: "SEARCH_BOOKMARK_SUGGESTION_WINDOW",

    CREATE_NEW_BOOKMARK: "CREATE_NEW_BOOKMARK",
    CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW: "CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW",

    EDIT_BOOKMARK: "EDIT_BOOKMARK",
} as const;

type HOTKEY_SCOPES = (typeof HOTKEY_SCOPES)[keyof typeof HOTKEY_SCOPES]


export function useAppHotkey() {
    const hotkeysContext = useHotkeysContext()

    const switchScope = (to: HOTKEY_SCOPES) => {
        hotkeysContext.enabledScopes.forEach(scope => {
            hotkeysContext.disableScope(scope)
        })
        hotkeysContext.enableScope(to)
    }

    return {
        switchScope
    }
}