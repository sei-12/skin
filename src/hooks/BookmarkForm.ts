import { useCallback, useEffect, useRef } from "react";
import type { FindTagMethod } from "../components/SuggestionWindow";
import { HOTKEY_SCOPES, useAppHotkey } from "./hotkey";
import { useHotkeys } from "react-hotkeys-hook";
import { useTagInputBox } from "./TagInputBox";
import { DB } from "../services/database";
import { useConfig } from "../providers/configProvider";
import type { BookmarkFormProps } from "../components/BookmarkForm";
import { useNotice } from "../providers/NoticeProvider";

export function useBookmarkForm(
    onClickDone: () => void,
    onClickCancel: () => void,
    findTagMethod: FindTagMethod,
    onChangeUrl: (url: string) => void
) {
    const titleRef = useRef<HTMLInputElement>(null);
    const urlRef = useRef<HTMLInputElement>(null);
    const descRef = useRef<HTMLInputElement>(null);
    const { addNotice } = useNotice()

    const tagInputBoxHook = useTagInputBox(findTagMethod);

    const setContent = async (title: string, desc: string, tags: string[]) => {
        if (titleRef.current === null) {
            return;
        }
        if (descRef.current === null) {
            return;
        }
        titleRef.current.value = title;
        descRef.current.value = desc;

        const inputedTags = await Promise.all(
            tags.map(t => DB.isExistsTag(t)
                .then(a => ({ text: t, exists: a }))
                .catch(() => {
                    addNotice({
                        message: "ERROR!",
                        serverity: "error"
                    })
                    return { text: "ERROR", exists: false }
                })
            )
        )
        tagInputBoxHook.setInputedTags(inputedTags)
    };

    const setUrl = (url: string) => {
        if (urlRef.current === null) {
            return;
        }
        urlRef.current.value = url;
        // イベントを発生させたかったけど、少し難しかった。
        onChangeUrl(url);
    };

    const getInputData = useCallback(() => {
        let title = ""
        let url = ""
        let desc = ""

        if (titleRef.current !== null) {
            title = titleRef.current.value
        }
        if (urlRef.current !== null) {
            url = urlRef.current.value
        }
        if (descRef.current !== null) {
            desc = descRef.current.value
        }

        return {
            title,
            desc,
            url,
            tags: tagInputBoxHook.inputedTags.map((e) => e.text),
        };
    }, [tagInputBoxHook.inputedTags]);

    const takeInputTag = useCallback(async () => {
        const inputBox =
            tagInputBoxHook.inputBoxRef.current;
        if (inputBox === null) {
            return;
        }

        const item = inputBox.value.replace(/ /g, "");

        if (item === "") {
            return;
        }

        const inputedTags = tagInputBoxHook.inputedTags.map(e => e.text)

        const has = inputedTags.find((t) => t == item) !== undefined;
        if (has) {
            inputBox.value = ""
            return;
        }

        const exists = await DB.isExistsTag(item).catch(() => {
            addNotice({
                message: "ERROR!",
                serverity: "error"
            })
            return false;
        });

        tagInputBoxHook.setInputedTags((ary) => {
            return [...ary, { text: item, exists }];
        });
        tagInputBoxHook.suggestionWindowHook.close();
        inputBox.value = "";
    }, [tagInputBoxHook.inputedTags])

    const onKeyDownBackspace = useCallback(() => {
        const hasFocus = tagInputBoxHook.inputBoxRef.current == document.activeElement
        if (!hasFocus) { return }
        tagInputBoxHook.popInputedTag()
    }, [tagInputBoxHook])

    const clearData = () => {
        if (titleRef.current === null) {
            return;
        }
        if (urlRef.current === null) {
            return;
        }
        if (descRef.current === null) {
            return;
        }

        titleRef.current.value = "";
        descRef.current.value = "";
        urlRef.current.value = "";

        tagInputBoxHook.setInputedTags([]);
    };

    const appHotkeyHook = useAppHotkey();
    const { keybinds } = useConfig()

    useEffect(() => {
        if (
            tagInputBoxHook.suggestionWindowHook.items
                .length === 0
        ) {
            appHotkeyHook.switchScope(HOTKEY_SCOPES.CREATE_NEW_BOOKMARK);
        } else {
            appHotkeyHook.switchScope(
                HOTKEY_SCOPES.CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW
            );
        }
    }, [tagInputBoxHook.suggestionWindowHook.items]);

    useHotkeys(
        keybinds.cancelCreateNewBookmark,
        onClickCancel,
        {
            scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK],
            preventDefault: true,
            enableOnFormTags: true,
        },
    );

    useHotkeys(
        keybinds.doneCreateNewBookmark,
        onClickDone,
        {
            scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK],
            preventDefault: true,
            enableOnFormTags: true,
        },
    );

    useHotkeys(
        keybinds.addFocusedSuggestionItem,
        tagInputBoxHook.addFocusedSuggestionItem,
        {
            scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW],
            preventDefault: true,
            enableOnFormTags: true,
        },
    );

    useHotkeys(
        keybinds.popInputedTag,
        onKeyDownBackspace,
        { scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK], enableOnFormTags: true },
    );


    useHotkeys("/", tagInputBoxHook.focusPredicateInputBox,
        {
            scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK],
            preventDefault: true,
        }
    )

    useHotkeys(
        keybinds.takeInputTag,
        takeInputTag,
        {
            scopes: [
                HOTKEY_SCOPES.CREATE_NEW_BOOKMARK,
                HOTKEY_SCOPES.CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW,
            ],
            keyup: true,
            keydown: false,
            enableOnFormTags: true,
        },
    );

    useHotkeys(
        keybinds.focusDownSuggestionWindow,
        tagInputBoxHook.suggestionWindowHook.focusDown,
        {
            scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW],
            preventDefault: true,
            enableOnFormTags: true,
        },
    );

    useHotkeys(
        keybinds.focusUpSuggestionWindow,
        tagInputBoxHook.suggestionWindowHook.focusUp,
        {
            scopes: [HOTKEY_SCOPES.CREATE_NEW_BOOKMARK_SUGGESTION_WINDOW],
            preventDefault: true,
            enableOnFormTags: true,
        },
    );

    const { colorTheme } = useConfig()
    const props: BookmarkFormProps = {
        titleRef,
        descRef,
        urlRef,
        tagInputBox: tagInputBoxHook.props,
        onClickCancel,
        onClickDone,
        onChangeUrl,
        colorTheme,
    }

    return {
        props,
        tagInputBoxHook,
        onKeyDownBackspace,
        setContent,
        getInputData,
        clearData,
        takeInputTag,
        setUrl,
    };
}