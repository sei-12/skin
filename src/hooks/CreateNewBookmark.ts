import { useCallback, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { invoke } from "@tauri-apps/api/core";
import { findTagMethod } from "../services/findTagMethod";
import { DB } from "../services/database";
import { isUrl } from "../vanilla/isUrl";
import { ClipBoardManager } from "../services/clipboard";
import { useBookmarkForm } from "./BookmarkForm";


export function useCreateNewBookmarkPage() {
    const navigate = useNavigate();

    const onClickCreateDone = () => {
        const inputData = bookmarkFormHook.getInputData();
        if (inputData === undefined) {
            return;
        }

        DB
            .insertBookmark(
                inputData.title,
                inputData.url,
                inputData.desc,
                inputData.tags
            )
            .then(() => {
                onClickCreateCancel();
            });
    };

    const onClickCreateCancel = () => {
        bookmarkFormHook.clearData();
        navigate("/");
    };

    const onChangeUrlInputBox = async (url: string) => {
        const content = (await invoke("fetch_website_content", { url })) as {
            title: string;
            desc: string;
        };
        const currentContent = bookmarkFormHook.getInputData();

        const setData = (cur: string | undefined, newContent: string | null) => {
            if (cur !== undefined && cur !== "") {
                return cur;
            }

            if (newContent !== null) {
                return newContent;
            }

            return "";
        };

        const title = setData(currentContent?.title, content.title);
        const desc = setData(currentContent?.desc, content.desc);
        bookmarkFormHook.setContent(title, desc);
    };

    const bookmarkFormHook = useBookmarkForm(
        onClickCreateDone,
        onClickCreateCancel,
        findTagMethod,
        onChangeUrlInputBox
    );

    const autoInputUrl = useCallback(async () => {
        const clipboardText = await ClipBoardManager.read()
        if (!isUrl(clipboardText)) {
            return;
        }
        bookmarkFormHook.setUrl(clipboardText)
    }, [])

    useEffect(() => {
        autoInputUrl()
    }, [])

    return bookmarkFormHook.props
}