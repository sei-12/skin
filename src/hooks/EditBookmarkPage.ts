import { useEffect } from "react";
import type { BookmarkFormProps } from "../components/BookmarkForm";
import { useLocation, useNavigate } from "react-router-dom";
import { useBookmarkForm } from "./BookmarkForm";
import { findTagMethod } from "../services/findTagMethod";
import { DB } from "../services/database";
import { useNotice } from "../providers/NoticeProvider";

interface State {
    bookmarkId: number;
}

export function useEditBookmarkPage(): BookmarkFormProps {
    const navigate = useNavigate();
    const { addNotice } = useNotice();
    const location = useLocation();
    const { bookmarkId } = location.state as State;

    const onClickDone = async () => {
        const inputedData = bookmarkFormHook.getInputData();
        const success = await DB.editBookmark(
            bookmarkId,
            inputedData.title,
            inputedData.url,
            inputedData.desc,
            inputedData.tags,
        )
            .then(() => {
                return true;
            })
            .catch(() => {
                return false;
            });

        if (success) {
            addNotice({
                message: "SUCCESS!",
                serverity: "success",
            });
            goRoot();
        } else {
            addNotice({
                message: "ERROR!",
                serverity: "error",
            });
        }
    };

    const goRoot = () => {
        navigate("/");
        bookmarkFormHook.clearData();
    };

    const bookmarkFormHook = useBookmarkForm(
        onClickDone,
        goRoot,
        findTagMethod,
        () => {},
    );

    const loadBookmarkData = async () => {
        const bkmk = await DB.getBookmark(bookmarkId);
        bookmarkFormHook.setContent(bkmk.title, bkmk.desc, bkmk.tags);
        bookmarkFormHook.setUrl(bkmk.url);
    };

    useEffect(() => {
        loadBookmarkData();
    }, []);

    return bookmarkFormHook.props;
}
