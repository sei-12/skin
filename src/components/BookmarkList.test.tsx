import { act, render, renderHook, screen } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { describe, expect, it, vi } from "vitest";
import { BookmarkList } from "./BookmarkList";
import { useBookmarkList } from "../hooks/BookmarkList";

describe("BookmarkList.tsx", () => {
    it("case1", async () => {
        // https://stackoverflow.com/questions/53271193/typeerror-scrollintoview-is-not-a-function
        window.HTMLElement.prototype.scrollIntoView = function() {};

        const onClickRemoveMock = vi.fn();
        const onClickCancelMock = vi.fn();

        const hook = renderHook(() =>
            useBookmarkList(onClickRemoveMock, onClickCancelMock)
        );

        act(() => {
            hook.result.current.setItems(items);
        });

        const result = render(
            <BookmarkList {...hook.result.current.props}></BookmarkList>
        );

        const bkmkItems = screen.getAllByTestId("bkmkitem");
        expect(bkmkItems.length).toBe(items.length);

        expect(result.asFragment()).toMatchSnapshot();
        

        items.forEach(item => {
            expect(screen.getByText(item.title)).toBeInTheDocument();
            expect(screen.getByText(item.desc)).toBeInTheDocument();
            item.tags.forEach(tag => {
                expect(screen.getByText(`#${tag}`)).toBeInTheDocument();
            });
        })
    });
});

const items = [
    {
        id: 1,
        title: "title1",
        url: "url1",
        desc: "desc1",
        tags: ["tag1", "tag2"],
    },
    {
        id: 2,
        title: "title2",
        url: "url2",
        desc: "desc2",
        tags: ["tag3", "tag4"],
    },
    {
        id: 3,
        title: "title3",
        url: "url3",
        desc: "desc3",
        tags: ["tag5", "tag6"],
    },
];
