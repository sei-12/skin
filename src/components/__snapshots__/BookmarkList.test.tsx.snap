// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookmarkList.tsx > case1 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-7jr0kh-MuiStack-root"
  >
    <div
      data-testid="bkmkitem"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-e5rvmw-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root css-13dhnfz-MuiCardContent-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1trx6ls-MuiTypography-root"
          >
            title1
          </h5>
          <div
            class="MuiStack-root css-1wcha3e-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1vt5eaj-MuiTypography-root"
            >
              #tag1
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1vt5eaj-MuiTypography-root"
            >
              #tag2
            </p>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-16rbfmb-MuiTypography-root"
          >
            desc1
          </p>
          <div
            class="MuiBox-root css-14s011l"
          >
            <div
              class="MuiStack-root css-iq4xaw-MuiStack-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="EditIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="bkmkitem"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-skzijy-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root css-13dhnfz-MuiCardContent-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1trx6ls-MuiTypography-root"
          >
            title2
          </h5>
          <div
            class="MuiStack-root css-1wcha3e-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1vt5eaj-MuiTypography-root"
            >
              #tag3
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1vt5eaj-MuiTypography-root"
            >
              #tag4
            </p>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-16rbfmb-MuiTypography-root"
          >
            desc2
          </p>
          <div
            class="MuiBox-root css-14s011l"
          >
            <div
              class="MuiStack-root css-iq4xaw-MuiStack-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="EditIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="bkmkitem"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-skzijy-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root css-13dhnfz-MuiCardContent-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1trx6ls-MuiTypography-root"
          >
            title3
          </h5>
          <div
            class="MuiStack-root css-1wcha3e-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1vt5eaj-MuiTypography-root"
            >
              #tag5
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1vt5eaj-MuiTypography-root"
            >
              #tag6
            </p>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-16rbfmb-MuiTypography-root"
          >
            desc3
          </p>
          <div
            class="MuiBox-root css-14s011l"
          >
            <div
              class="MuiStack-root css-iq4xaw-MuiStack-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="EditIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
